File: python/WebSocketService.py
import base64
import json
import requests
import threading
from websocket import WebSocketApp

class WebSocketService:
    def __init__(self, account_id, client_id, client_secret, url):
        self.url = url
        self.account_id = account_id
        self.client_id = client_id
        self.client_secret = client_secret
        self.oauth_url = "https://zoom.us/oauth/"
        self.ws = None
        self.heartbeat_thread = None

    def get_access_token(self):
        try:
            oauth_token = base64.b64encode(f"{self.client_id}:{self.client_secret}".encode()).decode()
            headers = {"Authorization": f"Basic {oauth_token}"}
            response = requests.post(
                f"{self.oauth_url}token?grant_type=account_credentials&account_id={self.account_id}",
                headers=headers
            )
            response.raise_for_status()
            data = response.json()
            return data.get("access_token")
        except Exception as e:
            print(f"{e}\nUnable to get access token. Check credentials.")
            return None

    def send_heartbeat(self):
        if self.ws:
            self.ws.send(json.dumps({"module": "heartbeat"}))
        self.heartbeat_thread = threading.Timer(30, self.send_heartbeat)
        self.heartbeat_thread.start()

    def on_open(self, ws):
        print("Connected")
        self.send_heartbeat()

    def on_close(self, ws, close_status_code, close_msg):
        print("Connection closed")
        if self.heartbeat_thread:
            self.heartbeat_thread.cancel()

    def on_message(self, ws, message):
        try:
            data = json.loads(message)
        except json.JSONDecodeError:
            print("Invalid JSON received")
            return

        if data.get("module") == "message":
            print(json.dumps(data.get("content"), indent=4))
            if data.get("content", {}).get("event") == "user.created":
                self.new_user_created_handler()

    def new_user_created_handler(self):
        print("\n\nA new user was created")
        print("Do some processing\n\n")

    def connect(self):
        try:
            access_token = self.get_access_token()
            if not access_token:
                raise Exception("Unable to get access token")

            self.ws = WebSocketApp(
                f"{self.url}&access_token={access_token}",
                on_open=self.on_open,
                on_close=self.on_close,
                on_message=self.on_message
            )
            self.ws.run_forever()
        except Exception as e:
            print(e)

File: javascript/WebSocketService.mjs
import WebSocket from "ws";
import { createHmac } from "node:crypto";

export class WebSocketService {
  constructor(accountId, clientId, clientSecret, url) {
    this.url = url;
    this.accountId = accountId;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.oauthUrl = "https://zoom.us/oauth/";
    this.ws = null;
    this.heartbeat = null;
  }

  async getAccessToken() {
    try {
      const oauthToken = Buffer.from(
        `${this.clientId}:${this.clientSecret}`
      ).toString("base64");

      const res = await fetch(
        `${this.oauthUrl}token?grant_type=account_credentials&account_id=${this.accountId}`,
        {
          method: "POST",
          headers: {
            Authorization: `Basic ${oauthToken}`,
          },
        }
      );

      const data = await res.json();
      return data.access_token;
    } catch (e) {
      console.error(
        `${e.message}\nUnable to get access token. Check credentials.`
      );
    }
  }

  sendHeartbeat() {
    this.ws.send(JSON.stringify({ module: "heartbeat" }));
    this.heartbeat = setTimeout(this.sendHeartbeat, 30000 * Math.random());
  }

  open() {
    console.log("connected");
    this.sendHeartbeat;
  }

  close() {
    console.log("connection closed");
    clearTimeout(this.heartbeat);
    this.ws = null;
  }

  async message(message) {
    let data = {};

    try {
      data = JSON.parse(message);
    } catch (e) {
      console.error("Invalid JSON received");
      return;
    }

    if (data?.module === "message") {
      console.log(JSON.parse(data.content));
      if (data?.content?.event === "user.created") this.newUserCreatedHandler;
    }
  }

  newUserCreatedHandler() {
    console.log("\n\n\n\nA new user was created");
    console.log("Do some processing\n\n\n\n");
  }

  async connect() {
    try {
      const accessToken = await this.getAccessToken();
      if (!accessToken) throw new Error("Unable to get access token");
      this.ws = new WebSocket(`${this.url}&access_token=${accessToken}`);
      this.ws.on("open", this.open);
      this.ws.on("close", this.close);
      this.ws.on("message", this.message);
    } catch (e) {
      console.error(e);
    }
  }
}
